README.md
Aurelia custom attribute to add awesome spinners based on http://tobiasahlin.com/spinkit/


**demo**
http://aurelia-spinner.azurewebsites.net/

**Installation**

```npm install aurelia-spinner --save```

**Plugin Setup**
The only required attribute its the spinner view.

```
import { SpinnerConfig, spinnerViews } from 'aurelia-spinner';

const spinnerConfig: SpinnerConfig = {
    spinner: spinnerViews.circle,
    useBackgroundOverlay: false,
    blockerClass: 'my-default-overlay-background',
    color: 'black'
};
 
aurelia.use.plugin(PLATFORM.moduleName('aurelia-spinner', spinnerConfig))

 ```

 **Configuration Options**
You can see all of the spinners available in the demo site.

 ```
 spinnerViews = {
  chasingDots
  circle,
  cubeGrid,
  doubleBounce,
  fadingCircle,
  foldingCube,
  pulse,
  rotatingPlane,
  threeBounce,
  wanderingCubes,
  wave
}
 ```

**add style in webpack**
if you use te spinner-block class you need to add this file.
```
'aurelia-spinner/dist/styles.css'

css: 'spinkit/css/spinkit.css;
sass: 'spinkit/scss/spinkit.scss;

or you can import a single spinner file under the spinner folder.

css: 'spinkit/css/spinners/9-cube-grid.css';
sass: 'spinkit/scss/spinners/9-cube-grid.scss';
```

**Usage**
Usage with default configurations
 ```
   
  <div class="container">
    <div spinner="show.bind: showVM"></div>
  </div>

  <div spinner="show.bind: showVM; view.bind: viewVM; block: true">
  </div>
  
  Currently, you need to wrap your custom element or code.
  You can specify the container element adding the spinner-container class first the children node. Also you can add the 'aurelia-spinner' class on the parent node to render the spinner component on top of your code.

  The background overlay class will be added on the container / wrapper node.

  <section spinner="show.bind: showVM">  
    <!-- spinner will be inserted here -->    
    <div class="spinner-container">  <!-- wrapper div -->  <!-- background ovelay will be added here -->
      <!-- my-custom-code -->
    </div>
  </section>

  <section class="aurelia-spinner">
  <!-- spinner will be inserted here -->     
    <div spinner="show.bind: showVM"> <!-- wrapper div -->  <!-- background ovelay will be added here -->   
    <!-- my-custom-code -->
    </div>
  </section>

  <section>
    <div spinner="show.bind: showVM" class="playground__spinner"> <!-- spinner-container class will be added here --> 
      <!-- spinner will be inserted here -->    
      stuff
    </div>
  </section>
  ```
  **Bindables**
```
  Bindable used to show or hide the spinner.
  show: boolean = false; 

  Just if you want to override the previous configured view.
  view: string = undefined;

  Used to display an overlay blocker with the spinner. You can change the style setting the class in the default configurarion object or overriding the css of the spinner-block class.
  block: boolean = false; 

  Type a color name or an hexa code.
  color: string = #333;

  example: 
  
  <div spinner="show.bind: showVM; block: true; view: viewVM; color: white"></div>

 ```

  **Dependencies**
  "spinkit": https://github.com/tobiasahlin/SpinKit


doc/CHANGELOG.md
<a name="1.0.10"></a>
## 1.0.10 (2017-09-04)

* adding color to spinner ([55f5b1d](https://github.com/ne0guille/aurelia-spinner/commit/55f5b1d))

<a name="1.0.9"></a>
## 1.0.9 (2017-08-31)

* adding plugin configuration -**not stable ([fc5a0a0](https://github.com/ne0guille/aurelia-spinner/commit/fc5a0a0))
* adding posibility to define spinner on parent node and blocker on elements first child ([10bbbe9](https://github.com/ne0guille/aurelia-spinner/commit/10bbbe9))
* adding spinner inside element ([8dbb188](https://github.com/ne0guille/aurelia-spinner/commit/8dbb188))
* adding tslint ([948a7c8](https://github.com/ne0guille/aurelia-spinner/commit/948a7c8))
* cmmit ([b4752fd](https://github.com/ne0guille/aurelia-spinner/commit/b4752fd))
* commit ([f98995c](https://github.com/ne0guille/aurelia-spinner/commit/f98995c))
* commit ([a203ba7](https://github.com/ne0guille/aurelia-spinner/commit/a203ba7))
* commit ([5a017e7](https://github.com/ne0guille/aurelia-spinner/commit/5a017e7))
* fixing block configuration ([e389351](https://github.com/ne0guille/aurelia-spinner/commit/e389351))
* fixing scripts ([3968f2b](https://github.com/ne0guille/aurelia-spinner/commit/3968f2b))
* fixing show llogical error ([cf2953e](https://github.com/ne0guille/aurelia-spinner/commit/cf2953e))
* fixing styling ([29753cf](https://github.com/ne0guille/aurelia-spinner/commit/29753cf))
* in development ([0e27b12](https://github.com/ne0guille/aurelia-spinner/commit/0e27b12))
* inital commit ([218dc97](https://github.com/ne0guille/aurelia-spinner/commit/218dc97))
* Initial commit ([07033dd](https://github.com/ne0guille/aurelia-spinner/commit/07033dd))
* prepare-release 1.0.2 ([c8f979a](https://github.com/ne0guille/aurelia-spinner/commit/c8f979a))
* preparing release ([e0700d4](https://github.com/ne0guille/aurelia-spinner/commit/e0700d4))
* preparing release 1.0.1 ([827eda2](https://github.com/ne0guille/aurelia-spinner/commit/827eda2))
* refacorting ([27232d6](https://github.com/ne0guille/aurelia-spinner/commit/27232d6))
* refacorting ([67d3068](https://github.com/ne0guille/aurelia-spinner/commit/67d3068))
* release 1.0.5 fixing block bindable configuration ([d29d7d7](https://github.com/ne0guille/aurelia-spinner/commit/d29d7d7))
* release v1.0.4 inserting spinner inside element's contianer ([68844b3](https://github.com/ne0guille/aurelia-spinner/commit/68844b3))
* releasing v1.0.3 ([bf146bf](https://github.com/ne0guille/aurelia-spinner/commit/bf146bf))
* separating logic in service class ([a031db5](https://github.com/ne0guille/aurelia-spinner/commit/a031db5))
* Update README.md ([266d27e](https://github.com/ne0guille/aurelia-spinner/commit/266d27e))
* Update README.md ([96abcd1](https://github.com/ne0guille/aurelia-spinner/commit/96abcd1))
* Update README.md ([c0acaab](https://github.com/ne0guille/aurelia-spinner/commit/c0acaab))
* Update README.md ([0e1d71a](https://github.com/ne0guille/aurelia-spinner/commit/0e1d71a))
* Update README.md ([21fc14a](https://github.com/ne0guille/aurelia-spinner/commit/21fc14a))



### 1.0.0-beta.1.1.0 (2016-03-23)

* Update to Babel 6

### 1.0.0-beta.1.0.0 (2016-01-29)


#### Features

* **all:** update jspm meta ([a2afb48e](http://github.com/aurelia/skeleton-plugin/commit/a2afb48e7cc83ab736b51ac2e4d5b7b3e3177ec2))


## 0.3.0 (2015-08-14)


#### Bug Fixes

* **index:** update to latest configuration api ([44f7a015](http://github.com/aurelia/skeleton-plugin/commit/44f7a015c0f15251bd07b327e42c875eaccbb735))


### 0.2.1 (2015-07-07)


## 0.2.0 (2015-06-08)


#### Bug Fixes

* **all:**
  * update to latest plugin api ([a050d736](http://github.com/aurelia/skeleton-plugin/commit/a050d736d32811066ffa902615cc73e1a5cbb6e3))
  * update compiler ([faf51acc](http://github.com/aurelia/skeleton-plugin/commit/faf51accc1514c6767eaed60df16dd3d586b5cc5))


#### Features

* **all:** initial commit of skeleton ([51a036d1](http://github.com/aurelia/skeleton-plugin/commit/51a036d146750a0bafd443dbc3def51ef7f89f6e))


test/unit/configure.spec.js
import {configure} from '../../src/index';

class ConfigStub {
  globalResources(...resources) {
    this.resources = resources;
  }
}

describe('the Aurelia configuration', () => {
  var mockedConfiguration;

  beforeEach(() => {
    mockedConfiguration = new ConfigStub();
    configure(mockedConfiguration);
  });

  it('should register a global resource', () => {
    expect(mockedConfiguration.resources).toContain('./spinner');
  });

});


test/unit/aurelia-spinner.spec.ts
import { bootstrap } from 'aurelia-bootstrapper';
import { Container, TaskQueue } from 'aurelia-framework';
import { PLATFORM } from 'aurelia-pal';
import { StageComponent } from 'aurelia-testing';

describe('testing spinner', () => {
  let component, taskQueue, isLoadingFalse;

  beforeEach(() => {
    isLoadingFalse = false;
    const view = `<div id="spinner-test" spinner="show.bind: isLoading" style="height: 500px">
                    <p>test</p>
                 </div>`;

    component = StageComponent
      .withResources(PLATFORM.moduleName('spinner'))
      .inView(view)
      .boundTo({ isLoading: isLoadingFalse });

    const container = new Container();
    taskQueue = container.get(TaskQueue);
  });

  it('should call createSpinner after attached', done => {
    let spy;

    component.manuallyHandleLifecycle().create(bootstrap)
      .then(() => spy = spyOn(component.viewModel, 'createSpinner'))
      .then(() => component.bind())
      .then(() => component.attached())
      .then(() => {
        expect(spy).toHaveBeenCalled();
        done();
      })
      .catch(e => console.log(e.toString()));
  });

  it('should toogle on show changed', done => {
    component.create(bootstrap).then(() => {
      const spy = spyOn(component.viewModel, 'showChanged')
        .and.callFake(() => component.viewModel.show = !isLoadingFalse);

      component.viewModel.showChanged();

      taskQueue.queueTask(() => {
        expect(component.viewModel.show).toBeTruthy();
        expect(spy).toHaveBeenCalled();
        done();
      });
    }).catch(e => console.log(e.toString()));
  });

  afterEach(() => component.dispose());

});


config.js
System.config({
  defaultJSExtensions: true,
  transpiler: false,
  paths: {
    "aurelia-spinner/*": "dist\\system/*.js",
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "aurelia-framework": "npm:aurelia-framework@1.1.4",
    "aurelia-pal": "npm:aurelia-pal@1.3.0",
    "aurelia-polyfills": "npm:aurelia-polyfills@1.0.0",
    "split.js": "npm:split.js@1.3.5",
    "npm:aurelia-binding@1.2.1": {
      "aurelia-logging": "npm:aurelia-logging@1.3.1",
      "aurelia-metadata": "npm:aurelia-metadata@1.0.3",
      "aurelia-pal": "npm:aurelia-pal@1.3.0",
      "aurelia-task-queue": "npm:aurelia-task-queue@1.2.0"
    },
    "npm:aurelia-dependency-injection@1.3.1": {
      "aurelia-metadata": "npm:aurelia-metadata@1.0.3",
      "aurelia-pal": "npm:aurelia-pal@1.3.0"
    },
    "npm:aurelia-framework@1.1.4": {
      "aurelia-binding": "npm:aurelia-binding@1.2.1",
      "aurelia-dependency-injection": "npm:aurelia-dependency-injection@1.3.1",
      "aurelia-loader": "npm:aurelia-loader@1.0.0",
      "aurelia-logging": "npm:aurelia-logging@1.3.1",
      "aurelia-metadata": "npm:aurelia-metadata@1.0.3",
      "aurelia-pal": "npm:aurelia-pal@1.3.0",
      "aurelia-path": "npm:aurelia-path@1.1.1",
      "aurelia-task-queue": "npm:aurelia-task-queue@1.2.0",
      "aurelia-templating": "npm:aurelia-templating@1.4.2"
    },
    "npm:aurelia-loader@1.0.0": {
      "aurelia-metadata": "npm:aurelia-metadata@1.0.3",
      "aurelia-path": "npm:aurelia-path@1.1.1"
    },
    "npm:aurelia-metadata@1.0.3": {
      "aurelia-pal": "npm:aurelia-pal@1.3.0"
    },
    "npm:aurelia-polyfills@1.0.0": {
      "aurelia-pal": "npm:aurelia-pal@1.3.0"
    },
    "npm:aurelia-task-queue@1.2.0": {
      "aurelia-pal": "npm:aurelia-pal@1.3.0"
    },
    "npm:aurelia-templating@1.4.2": {
      "aurelia-binding": "npm:aurelia-binding@1.2.1",
      "aurelia-dependency-injection": "npm:aurelia-dependency-injection@1.3.1",
      "aurelia-loader": "npm:aurelia-loader@1.0.0",
      "aurelia-logging": "npm:aurelia-logging@1.3.1",
      "aurelia-metadata": "npm:aurelia-metadata@1.0.3",
      "aurelia-pal": "npm:aurelia-pal@1.3.0",
      "aurelia-path": "npm:aurelia-path@1.1.1",
      "aurelia-task-queue": "npm:aurelia-task-queue@1.2.0"
    }
  }
});


src/spinner.ts
import { bindable, bindingMode, inject, TaskQueue } from 'aurelia-framework';

import { SpinnerService } from './spinner-service';

@inject(Element, TaskQueue, SpinnerService)
export class SpinnerCustomAttribute {
  private target: Element;

  @bindable({ defaultBindingMode: bindingMode.oneWay }) show: boolean = false;
  @bindable({ defaultBindingMode: bindingMode.oneTime }) view?: string = undefined;
  @bindable({ defaultBindingMode: bindingMode.oneTime }) block: boolean = false;
  @bindable({ defaultBindingMode: bindingMode.oneTime }) color?: string = undefined;

  constructor(private element: Element,
    private taskQueue: TaskQueue,
    private spinnerService: SpinnerService) { }

  bind() {
    this.view = this.view || this.spinnerService.config.spinner;
    this.block = !!(this.block || this.spinnerService.config.useBackgroundOverlay);
    this.color = this.color || this.spinnerService.config.color;

    if (!this.view) throw new Error('no view has been specified for the spinner');
  }

  attached() {
    this.taskQueue.queueTask(async () => this.target = await this.spinnerService.createSpinner(this.element, this));
}

showChanged(newValue: boolean) {
  if (!this.target || !this.block) return;

  this.spinnerService.toogleBackgroundOverlay(this.target, newValue);
}
}


src/spinner-config.ts
import { PLATFORM } from 'aurelia-framework';

const blockerClass: string = 'spinner-blocker';

export interface SpinnerConfig {
  spinner?: string;
  color?: string;
  useBackgroundOverlay?: boolean;
  blockerClass?: string;
}

export class DefaultSpinnerConfig implements SpinnerConfig {
  spinner?: string = undefined;
  color?: string = '#333';
  useBackgroundOverlay: boolean = false;
  blockerClass: string = blockerClass;
}

export const spinnerViews = {
  chasingDots: PLATFORM.moduleName('aurelia-spinner/dist/views/chasing-dots.html'),
  circle: PLATFORM.moduleName('aurelia-spinner/dist/views/circle.html'), // TODO color
  cubeGrid: PLATFORM.moduleName('aurelia-spinner/dist/views/cube-grid.html'),
  doubleBounce: PLATFORM.moduleName('aurelia-spinner/dist/views/double-bounce.html'),
  fadingCircle: PLATFORM.moduleName('aurelia-spinner/dist/views/fading-circle.html'), // TODO color
  foldingCube: PLATFORM.moduleName('aurelia-spinner/dist/views/folding-cube.html'), // TODO color
  pulse: PLATFORM.moduleName('aurelia-spinner/dist/views/pulse.html'),
  rotatingPlane: PLATFORM.moduleName('aurelia-spinner/dist/views/rotating-plane.html'),
  threeBounce: PLATFORM.moduleName('aurelia-spinner/dist/views/three-bounce.html'),
  wanderingCubes: PLATFORM.moduleName('aurelia-spinner/dist/views/wandering-cubes.html'),
  wave: PLATFORM.moduleName('aurelia-spinner/dist/views/wave.html'),
};


src/aurelia-spinner.ts
import { autoinject, bindable, bindingMode, Container, View, ViewEngine, computedFrom } from 'aurelia-framework';
import { DOM } from 'aurelia-pal';

import { SpinnerConfig, spinnerView } from "./spinner-config";

@autoinject
export class SpinnerCustomAttribute {
  private divElement: HTMLElement;
  private target: Element;
  private readonly spinnerHtml: string = './aurelia-spinner.html';
  private config: SpinnerConfig;
  private _view: string;

  @bindable({ defaultBindingMode: bindingMode.oneWay }) show: boolean = false;
  @bindable({ defaultBindingMode: bindingMode.oneTime }) view: string = undefined;
  @bindable({ defaultBindingMode: bindingMode.oneTime }) block: boolean = false;

  static defaultConfig: SpinnerConfig = {
    view: "src/views/circle.html", //spinnerView.circle
    useBackgroundBlocker: false,
    blockerClass: 'spinner-blocker'
  };

  constructor(private element: Element,
    private container: Container,
    private viewEngine: ViewEngine,
    private spinnerConfig: 'spinner-config') {
    this.config = Object.assign({}, SpinnerCustomAttribute.defaultConfig);
    console.log(this.config);
    console.log(this.config.blockerClass);
    console.log(this.spinnerConfig);
  }

  bind() {
    this.view = this.view || this.config.view;
    this.block = this.block || this.config.useBackgroundBlocker;

    if (!this.view) throw new Error("no view has been specified for the spinner");
  }

  attached() {
    this.createSpinner();
  }

  showChanged(showSpinner: boolean) {
    if (!this.target && !this.block) return;

    showSpinner ? this.target.classList.add(this.config.blockerClass) :
      this.target.classList.remove(this.config.blockerClass);
  }

  viewChanged(newValue: string) {
    if(!newValue) this._view = this.config.view;
    
    console.log(this.view);
    console.log(spinnerView[newValue]);
    console.log(newValue);
    
    this._view = spinnerView[newValue];
  }

  private createSpinner(): void {
    this.viewEngine.loadViewFactory(this.spinnerHtml).then(factory => {
      const childContainer = this.container.createChild();
      const view = factory.create(childContainer);
      view.bind(this);

      this.addElement(view);
    });
  }

  private removeSpinner(): void {
    const removeSpinner = DOM.querySelectorAll(`#${this.element.id}`)[0];
    removeSpinner.removeChild(this.divElement);
  }

  private addElement(view: View): void {
    const container = DOM.querySelectorAll(`#${this.element.id}`)[0];

    const spinnerDivElement = <HTMLElement>DOM.createElement('div');
    view.appendNodesTo(spinnerDivElement);

    this.target = this.element.firstElementChild || this.element;

    this.divElement = this.setElementStyle(this.element, spinnerDivElement);

    if (this.block) this.target.classList.add(this.config.blockerClass);

    container.insertBefore(this.divElement, container.firstChild);
  }

  // todo fix too recalculate element height
  private setElementStyle(element: Element, htmlElement: HTMLElement): HTMLElement {
    const elementRect = element.getBoundingClientRect();
    const height = htmlElement.getBoundingClientRect().height;
    let top = elementRect.top + elementRect.height;

    top = ((top + height) < window.innerHeight) ? top + window.scrollY : (elementRect.top - height + window.scrollY);

    htmlElement.style.position = 'absolute';
    htmlElement.style.zIndex = '999';
    htmlElement.style.height = `calc(100% - ${top}px)`;
    htmlElement.style.top = `${top}px`;
    htmlElement.style.left = `calc(50% - 35px)`;

    return htmlElement;
  }

}


src/spinner-service.ts
import { Container, inject, View, ViewEngine, ViewFactory } from 'aurelia-framework';
import { PLATFORM } from 'aurelia-pal';

import { DefaultSpinnerConfig, SpinnerConfig } from './spinner-config';

const containerClass: string = 'spinner-container';
const spinnerClass: string = 'aurelia-spinner';
const spinnerHeight: number = 35;

@inject(Container, ViewEngine, 'spinner-config')
export class SpinnerService {
  private readonly spinnerHtml: string = PLATFORM.moduleName('aurelia-spinner/dist/views/spinner.html');
  public config: SpinnerConfig = new DefaultSpinnerConfig();

  constructor(private container: Container,
    private viewEngine: ViewEngine,
    private spinnerConfig: SpinnerConfig) {
    if (spinnerConfig) this.config = Object.freeze({ ...this.config, ...this.spinnerConfig });
  }

  public async createSpinner(element: Element, self: any): Promise<Element> {
    const factory: ViewFactory = await this.viewEngine.loadViewFactory(this.spinnerHtml);
    const childContainer: Container = this.container.createChild();
    const view: View = factory.create(childContainer);

    view.bind(self);

    let blockerContainer: Element = element;
    let spinnerContainer: Element = element;

    if (element.parentElement && this.containsClass(element.parentElement, spinnerClass))
      spinnerContainer = element.parentElement;

    if (element.firstElementChild && this.containsClass(element.firstElementChild, containerClass))
      blockerContainer = element.firstElementChild;

    element.classList.add(containerClass);

    this.addElement(spinnerContainer, view);

    if (this.config.useBackgroundOverlay || self.block) this.toogleBackgroundOverlay(blockerContainer, true);

    return blockerContainer;
  }

  public toogleBackgroundOverlay(target: Element, showSpinner: boolean): void {
    if (target && this.config.blockerClass)
      showSpinner ? target.classList.add(this.config.blockerClass) : target.classList.remove(this.config.blockerClass);
  }

  private addElement(spinnerContainer: Element, view: View): void {
    let spinnerDivElement: HTMLElement = <HTMLElement>document.createElement('div');
    view.appendNodesTo(spinnerDivElement);

    spinnerDivElement = this.setElementStyle(spinnerContainer, spinnerDivElement);

    spinnerContainer.appendChild(spinnerDivElement);
  }

  private setElementStyle(element: Element, htmlElement: HTMLElement): HTMLElement {
    const elementRect: ClientRect = element.getBoundingClientRect();
    const height: number = htmlElement.getBoundingClientRect().height || spinnerHeight;
    const top: string = (elementRect.top + height) > window.innerHeight ? `100px` : `calc(10% - ${spinnerHeight}px)`;

    htmlElement.style.position = 'absolute';
    htmlElement.style.zIndex = '999';
    htmlElement.style.top = top;
    htmlElement.style.left = `calc(50% - ${spinnerHeight}px)`;

    return htmlElement;
  }

  private containsClass(element: Element, className: string) {
    return element.classList.contains(className);
  }
}
